### DISCUSSION TITLE
The parallelism of OpenCL kernel when POCL_DEVICES="cpu"

### DISCUSSION THREAD

**Discussion by soccercheng** [pocl/pocl#1511] [Q&A] [answered]
Hi,

I am new to pocl and I've been tracing the calling sequence and parallelism architecture of pocl when POCL_DEVICES="cpu".

From my understanding is that pocl uses cland/llvm to front-end-ly compoile kernel.cl with the folling stages:

1. Compile kernel.cl to program program.bc with limited optimization
2. Optimize and parallelize program.bc to parallel.bc 
3. Generate object code for target CPU and link it to ".so" library

The followings are my furrther questions 

**[How pocl parallelly executes the kernel code]**

1. Is the ".so" itself embedded with parallel mechenison and spawn threads when running    or    it is pocl runtime that spawns threads and each thread uses the library to run the kernel parallelly.
2. How "data" is split to each thread ?  Does pocl adopt SPMD (simgle program/function, multiple data) architecture ?

**[Link library dependencies]**

The following is the library dependency of generated kernel.so, I am quirous about why the generated kernel.so needs to link against these libraries?

$ ldd ./kcache/.../softmax_with_chunked_sum_kernel/32-8-1-goffs0-smallgrid/softmax_with_chunked_sum_kernel.so
	linux-vdso.so.1 (0x00007fff9ade1000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007b9a5e19e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007b9a5de00000)
	/lib64/ld-linux-x86-64.so.2 (0x00007b9a5e2be000)

**Comment by franz** [comment, marked as answer]
(note: my answers are for CPU driver only)

>  Is the ".so" itself embedded with parallel mechenison and spawn threads when running 

No, the '.so' contains only a kernel compiled to a function that executes all workitems of a single workgroup. It is pretty much the `parallel.bc` file compiled to an object file; you can check (with `llvm-dis`) that there is no threading related code in `parallel.bc`.

> or it is pocl runtime that spawns threads and each thread uses the library to run the kernel parallelly.

Yes, the runtime decides how to run the workgroup functions. The "cpu" driver uses multiple threads by default, but it can be tweaked to use less.

> How "data" is split to each thread ? Does pocl adopt SPMD (simgle program/function, multiple data) architecture ?

That depends on the workgroup "method" selected at kernel compilation time. Most often it is the `loops` method that creates a loop to execute all workitems of a workgroup, while making sure to correctly handle barriers etc. There are two more methods, `repl` that replicates the code of each workitem of a workgroup (can be useful for VLIW processors) and `cbs` (continuation-based synchronization) that is able to handle some corner cases that the `loops` can't handle.

> The following is the library dependency of generated kernel.so, I am quirous about why the generated kernel.so needs to link against these libraries?

1) depending on some cmake options, and possibly clang/llvm version, some of the OpenCL builtins in the user's OpenCL source code may end up compiled to a call to libm function. We've tried to avoid this as much as possible with the default cmake configuration (to make the computation independent of libm & compiler versions), but we don't check this regularly and we don't check all configurations.
2) depending on some cmake options, the OpenCL `printf` builtin function is (by default) partially implemented with `snprintf`, and may also require `write`, therefore we need libc. There is also an implementation that is completely self-contained (does not need snprintf on the kernel side, writes only into the printf buffer) but it's disabled by default, since the printed results (while correct) are not exactly the same as libc's snprintf.

**Comment by pjaaskel** [comment]
Michal's answer covers well except I'd add that we use 'loopvec' also to autovectorize across the work-groups' work-items by marking the loops with "parallel loop metadata" to make it easier to analyze for deps and passing them to the standard LLVM loop vectorizer(s). More reading: https://link.springer.com/article/10.1007/s10766-014-0320-y and https://portablecl.org/docs/html/kernel_compiler.html.

**Comment by soccercheng** [comment]
Many thanks for your valuable input. now I've basic understanding about, the CPU/pthread driver, the pocl runtime that is able to "parallelly" scheduling the execution of a kernel.so.

and.. how about the almaif driver ? 
How does it schedule the triggerinng an "AFOCLized" kernel on Xilinx FPGA, since there is only one single instance of  "AFOCLized" kernel on Xilinx FPGA, it seems NOT possible to parallelly schedule jobs onto a single AFOCLized" kernel instance ?

**Reply by pjaaskel** [reply]
I believe @TopiLeppanen answered you in email, but summarizing: The packet processor can schedule work to multiple hardware blocks if there are any, regardless if they implement the same function or not. I do not know if that is implemented in the currently contributed open source code or not.

**Reply by TopiLeppanen** [reply]
No, there are no such open-source examples atm. But should be pretty straight-forward to implement, as the packet processor can execute the entire built-in kernel however it likes.

**Reply by TopiLeppanen** [reply]
You asked in an email a follow-up question to this:
"I am also curious if the current pocl almaif scheduler/driver is able to distribute workitems to  IP instances of the same identical kernel accordingly."
I will answer here, to move the discussion to the public place where it may be of use to somebody else.

Basically there is currently no 'scheduler' in the almaif driver to do any of that. The entire kernel dispatch packet gets pushed to the AlmaIF device via the AQL queue. Then the packet/command processor core must execute it in its entirety. If one wanted to add some smart scheduling/distribution logic, it would be added there, to the command processor firmware. 

For example, in a hypothetical system if there were 4 identical accelerator IP instances and a single command processor core on an FPGA, the firmware would have to be modified to adjust the argument pointers it uses for each of the IP instances so that they operate on different data. For built-in kernels this is quite doable, as you always know the access pattern well ahead of time. Then, the command processor core would wait until every IP instance is done with their computation, and only then mark the entire dispatch command completed. In this sense, the AlmaIF driver is completely oblivious of any splitting.
